# Εισαγωγή και επισκόπηση
The dataset concerns with the classification of two mental stages from recorded EEG signals: Planning (during imagination of motor act) and Relax state.

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('../dataset/plrx.txt', delimiter = "\t", header=None)
df

0         1         2         3         4         5         6   \
0   -0.17936 -0.207000 -0.209710 -0.097260 -0.119210 -0.173220 -0.280760   
1   -0.14659 -0.164940  0.249870 -0.073985  0.494940 -0.156330 -0.288910   
2   -0.13131  0.518160  0.432810  0.397890  0.043127  0.563200 -0.013246   
3    0.64940  0.418780  0.148350  0.333490 -0.219740  0.398910  0.590650   
4   -0.40628  0.115030  0.073336  0.070066 -0.030920  0.152050 -0.191420   
..       ...       ...       ...       ...       ...       ...       ...   
177 -0.55005  0.276290 -0.273240  0.113190 -0.776980  0.522660 -0.850620   
178 -0.78399  0.254600 -0.700990 -0.299170 -0.858330  0.210500 -0.530050   
179  0.29877 -0.179620  0.032859 -0.899010  0.540300 -0.361840  0.177900   
180 -0.37812 -0.522870  0.215410 -0.243590  0.637840 -0.492210 -0.506930   
181  0.37913 -0.046562 -0.123360 -0.063958 -0.109830 -0.045824  0.123800   

           7         8         9         10        11   12  13  
0    0.223170  0.418660 -0.032886  0.003383 -0.334250  1.0 NaN  
1    0.500800 -0.045553  0.050759 -0.017511  0.066959  1.0 NaN  
2   -0.062833 -0.883390 -0.040303  0.088057  0.631200  2.0 NaN  
3    0.077100 -1.107200 -0.367230  0.364680  0.571720  1.0 NaN  
4   -0.580430 -0.102250  0.153750 -0.126950  0.184570  1.0 NaN  
..        ...       ...       ...       ...       ...  ...  ..  
177  0.440040  0.203910 -0.014462  0.002273 -0.100480  2.0 NaN  
178  0.518550 -0.303190  0.020018 -0.020915  0.663940  1.0 NaN  
179  0.713250 -0.098148 -0.411130  0.246830 -0.389160  2.0 NaN  
180 -0.258110 -0.328990 -0.039117  0.007591  0.476030  1.0 NaN  
181  0.496240 -0.641570 -0.242290  0.173690  0.490630  2.0 NaN  

[182 rows x 14 columns]

In [3]:
df.shape
#τα δείγματα (instances) είναι οι γραμμές και τα χαρακτηριστικά (attributes) οι στήλες
#χρειάστηκε το header=None για να μην παίρνει ως επικεφαλίδα την πρώτη γραμμή

#λόγω space στο txt έχει μια έξτρα στήλη NaN που πρέπει να διορθωθεί.
#df.shape -> διαστάσεις dataframe

(182, 14)

In [4]:
df.isnull()

#εντολή isnull δίνεται η τιμή True όπου υπάρχει κενή τιμή στο dataset, με την isnull.sum() μπορούμε να δούμε σε κάθε κολόνα
#πόσα κενά υπάρχουν

pd.isnull(df).sum()

#παρατηρούμε οτι στην κολόνα 13 λείπουν όλες οι τιμές 

percent_missing = df.isnull().sum() * 100 / len(df)
percent_missing



0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
5       0.0
6       0.0
7       0.0
8       0.0
9       0.0
10      0.0
11      0.0
12      0.0
13    100.0
dtype: float64

In [5]:
df.drop(13, axis=1, inplace=True)
#axis=1 όταν πρόκεται για κολόνα, inplace=True για να κάνει την αλλαγή και όχι μόνο να μας δείξει πως θα ήταν
df.shape

(182, 13)

In [6]:
df

0         1         2         3         4         5         6   \
0   -0.17936 -0.207000 -0.209710 -0.097260 -0.119210 -0.173220 -0.280760   
1   -0.14659 -0.164940  0.249870 -0.073985  0.494940 -0.156330 -0.288910   
2   -0.13131  0.518160  0.432810  0.397890  0.043127  0.563200 -0.013246   
3    0.64940  0.418780  0.148350  0.333490 -0.219740  0.398910  0.590650   
4   -0.40628  0.115030  0.073336  0.070066 -0.030920  0.152050 -0.191420   
..       ...       ...       ...       ...       ...       ...       ...   
177 -0.55005  0.276290 -0.273240  0.113190 -0.776980  0.522660 -0.850620   
178 -0.78399  0.254600 -0.700990 -0.299170 -0.858330  0.210500 -0.530050   
179  0.29877 -0.179620  0.032859 -0.899010  0.540300 -0.361840  0.177900   
180 -0.37812 -0.522870  0.215410 -0.243590  0.637840 -0.492210 -0.506930   
181  0.37913 -0.046562 -0.123360 -0.063958 -0.109830 -0.045824  0.123800   

           7         8         9         10        11   12  
0    0.223170  0.418660 -0.032886  0.003383 -0.334250  1.0  
1    0.500800 -0.045553  0.050759 -0.017511  0.066959  1.0  
2   -0.062833 -0.883390 -0.040303  0.088057  0.631200  2.0  
3    0.077100 -1.107200 -0.367230  0.364680  0.571720  1.0  
4   -0.580430 -0.102250  0.153750 -0.126950  0.184570  1.0  
..        ...       ...       ...       ...       ...  ...  
177  0.440040  0.203910 -0.014462  0.002273 -0.100480  2.0  
178  0.518550 -0.303190  0.020018 -0.020915  0.663940  1.0  
179  0.713250 -0.098148 -0.411130  0.246830 -0.389160  2.0  
180 -0.258110 -0.328990 -0.039117  0.007591  0.476030  1.0  
181  0.496240 -0.641570 -0.242290  0.173690  0.490630  2.0  

[182 rows x 13 columns]

In [7]:
df[12].unique() #οι κλάσεις είναι 2 με τιμές 1 και 2 όπως εκτυπώνει η εντολή , binary dataset

array([1., 2.])

In [8]:
#υπάρχουν ετικέτες κλάσεων (1,2) και βρίσκονται στην 13η κολόνα

#με την εντολή isnull δίνεται η τιμή True όπου υπάρχει κενή τιμή στο dataset, με την isnull.sum() μπορούμε να δούμε σε κάθε κολόνα
#πόσα κενά υπάρχουν

In [9]:
#με την παρακάτω εντολή μετράμε στην 13η κολόνα πόσες 1.0 τιμές υπάρχουν, άρα 130->1.0 και 52->2.0 άρα μη ισορροπημένο dataset
df[12].value_counts()[1.0]

130

In [10]:
from sklearn.model_selection import train_test_split
X = df.drop(12,axis=1)
y = df[12]
X_train, X_test, y_train, y_test= train_test_split(X, y, test_size=0.3, random_state=7)
#το random_state αν ορίσουμε έναν int τότε βγάζει κάθε φορά το ίδιο αποτέλεσμα

Παρακάτω βλέπουμε ότι δεν έχουμε διαφορετικούς τύπους στο dataset, μη κατηγορικά features

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 182 entries, 0 to 181
Data columns (total 13 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       182 non-null    float64
 1   1       182 non-null    float64
 2   2       182 non-null    float64
 3   3       182 non-null    float64
 4   4       182 non-null    float64
 5   5       182 non-null    float64
 6   6       182 non-null    float64
 7   7       182 non-null    float64
 8   8       182 non-null    float64
 9   9       182 non-null    float64
 10  10      182 non-null    float64
 11  11      182 non-null    float64
 12  12      182 non-null    float64
dtypes: float64(13)
memory usage: 18.6 KB


In [28]:
X_test

0         1         2         3         4         5         6   \
139  0.175930  0.010151 -0.130770  0.026617 -0.194560 -0.005084  0.128660   
36   0.490530 -0.352450 -0.208380 -0.124700  0.089263 -0.417290  0.370170   
111 -1.667000 -1.004900 -0.089985 -0.545560  0.634780 -0.845120 -1.651000   
119 -1.044000  0.112720  0.071678  0.350820 -0.099634  0.175110 -0.599540   
124  0.057557  0.066469  0.503080  0.321870  0.560450  0.020600  0.221450   
134  0.582170 -0.045015 -0.191870 -0.012795 -0.377450  0.042399  0.294060   
102  0.270760  0.453680 -0.296690 -0.214630 -0.601760  0.377740  0.364620   
32   0.020291 -0.075990 -0.630340 -0.496820 -0.441020 -0.229650  0.113990   
121  0.401600  0.520670 -0.343130  0.129090 -0.943360  0.597760  0.188090   
146 -0.457710 -0.240030 -0.337060 -0.312570 -0.321920 -0.107490 -0.613890   
22  -0.813330 -0.896250 -0.016994 -0.443280  0.658210 -0.763470 -0.964190   
116  0.451780  0.345550 -0.195670  0.196130 -0.568480  0.304380  0.456340   
164  0.264880 -0.233240 -0.219270 -0.447010  0.080071 -0.314160  0.121150   
120  0.111900  1.000000 -0.248050  0.103810 -1.077700  0.979000  0.273420   
85   0.700530 -0.245190  0.140190 -0.017159  0.473590 -0.390340  0.701160   
155  0.125360 -0.105870  0.297200  0.189690  0.385210 -0.075452  0.105350   
171 -0.146480 -0.104560  0.270550  0.423220  0.213750 -0.053184 -0.078576   
162 -0.313380 -0.003663  0.085749  0.071735  0.046584  0.040560 -0.342410   
51   0.974870  0.793910 -0.201030  0.346940 -0.903830  0.720920  1.000000   
45  -0.096201 -0.713850  0.286050 -0.189620  0.948600 -0.721870 -0.163160   
74  -0.232990 -0.127130  0.416250  0.198750  0.616670 -0.136940 -0.168230   
66  -0.151800  0.596520  0.148330  0.557390 -0.495660  0.705950  0.001053   
166  0.109950  0.043399 -0.407680 -0.479300 -0.300180 -0.052069  0.311020   
96  -0.754070 -0.838310 -0.664870 -0.961680  0.080456 -0.916360 -0.632160   
13  -0.659590  0.200720  0.404650  0.297540  0.236830  0.292860 -0.416750   
168  0.109950  0.043399 -0.407680 -0.479300 -0.300180 -0.052069  0.311020   
86   0.080397 -0.630740 -0.156610 -0.177550  0.318310 -0.627410 -0.157700   
180 -0.378120 -0.522870  0.215410 -0.243590  0.637840 -0.492210 -0.506930   
153 -0.301400 -0.133870 -0.296810 -0.522410 -0.058147 -0.262560  0.041177   
169 -0.638260  0.377300 -0.696330 -0.507110 -0.974550  0.361490 -0.246800   
129 -0.025852 -0.040066 -0.240340 -0.463700 -0.101370 -0.096433  0.005884   
24   0.355550  0.120360  0.035551  0.222860 -0.111990  0.099171  0.277400   
150  0.289660 -0.015666 -0.746660 -0.231950 -0.991140  0.019864  0.236960   
126 -0.086402 -0.114480 -0.161240 -0.558040  0.126710 -0.194420 -0.023637   
97  -0.188400 -0.547120 -0.070795 -0.295190  0.242760 -0.429230 -0.486230   
78  -0.340500 -0.344610  0.101650 -0.279840  0.437910 -0.308190 -0.465730   
128 -0.055702 -0.254080  0.003567 -0.204410  0.118580 -0.170600 -0.160740   
177 -0.550050  0.276290 -0.273240  0.113190 -0.776980  0.522660 -0.850620   
11   0.461080  0.465300  0.931170  0.889990  0.687710  0.455770  0.486050   
172  0.544040  0.209940  0.324380 -0.135440  0.377960  0.080315  0.773450   
27  -0.737400 -0.456400 -0.527740 -0.479010 -0.199990 -0.469120 -0.497730   
95   0.217900  0.071494  0.033231 -0.112710  0.073063  0.039341  0.032274   
113  0.497770 -0.186880  0.174200 -0.261790  0.562630 -0.326820  0.585290   
140  0.391610 -0.126570  0.166420  0.283960  0.129760 -0.058800  0.086254   
106 -0.515470  0.631850  0.263320  0.315880 -0.210860  0.687630 -0.233990   
37   0.329120 -0.147180 -0.386830 -0.243550 -0.282850 -0.210900  0.214710   
28   0.213290 -0.004894 -0.317520 -0.218750 -0.339420 -0.003355  0.027455   
63   0.184430 -0.519810  0.147670 -0.337440  0.686380 -0.619240  0.353310   
82  -0.139390  0.364710  0.124060  0.770890 -0.431670  0.522730 -0.310290   
65   1.000000  0.074734 -0.272100  0.044988 -0.337910 -0.022484  0.801690   
125 -0.107410 -0.242160  0.018824  0.164830  0.149920 -0.232660 -0.124740   
99  -0.

# Ταξινόμηση 

In [16]:
from sklearn.metrics import accuracy_score
from sklearn.metrics import f1_score

### Για τον ταξινομητή Dummy

In [13]:
from sklearn.dummy import DummyClassifier
dummy = DummyClassifier()
dummy.fit(X_train, y_train)

DummyClassifier()

In [17]:
y_pred = dummy.predict(X_test)
accuracy_score(y_test,y_pred)

0.7272727272727273

In [18]:
f1_score(y_test, y_pred, average='binary')

0.8421052631578948

### Για την μέθοδο Gaussian Naive Bayes (GNB)

In [19]:
import matplotlib.pyplot as plt
from sklearn.naive_bayes import GaussianNB
NBModel=GaussianNB()
NBModel.fit(X_train,y_train)

GaussianNB()

In [20]:
y_pred = NBModel.predict(X_test)
from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_pred)

0.5636363636363636

In [21]:
from sklearn.metrics import f1_score
f1_score(y_test, y_pred, average='binary')

0.7

### Για το KNN model

In [22]:
from sklearn.neighbors import KNeighborsClassifier
neigh = KNeighborsClassifier()
neigh.fit(X_train,y_train)

KNeighborsClassifier()

In [23]:
y_pred = neigh.predict(X_test)
accuracy_score(y_test, y_pred)

0.6545454545454545

In [24]:
f1_score(y_test, y_pred, average='binary')

0.7912087912087913

### Για το LR

In [25]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
lr.fit(X_train, y_train)

LogisticRegression()

In [26]:
y_pred = lr.predict(X_test)
accuracy_score(y_test, y_pred)

0.7090909090909091

In [27]:
f1_score(y_test, y_pred, average='binary')

0.8297872340425532

## Σχήμα διασταυρούμενης επικύρωσης

In [35]:
from sklearn.model_selection import cross_validate
scores_dummy = cross_validate(dummy, X_train, y_train, cv=10, scoring=['accuracy','f1_macro'])
scores_dummy

{'fit_time': array([0.00093603, 0.00073814, 0.00048018, 0.0004847 , 0.00047398,
        0.00052094, 0.00047612, 0.0004878 , 0.00051308, 0.00111532]),
 'score_time': array([0.00173903, 0.00115895, 0.00127602, 0.00125694, 0.00114584,
        0.00114918, 0.00124812, 0.00120926, 0.00158691, 0.00248694]),
 'test_accuracy': array([0.69230769, 0.69230769, 0.69230769, 0.69230769, 0.69230769,
        0.69230769, 0.69230769, 0.75      , 0.75      , 0.75      ]),
 'test_f1_macro': array([0.40909091, 0.40909091, 0.40909091, 0.40909091, 0.40909091,
        0.40909091, 0.40909091, 0.42857143, 0.42857143, 0.42857143])}

In [45]:
scores_nb = cross_validate(NBModel, X_train, y_train, cv=10, scoring=['accuracy','f1_macro'])
scores_nb

{'fit_time': array([0.00298381, 0.00204015, 0.00175309, 0.00169611, 0.00192714,
        0.0029068 , 0.00240803, 0.00177622, 0.00255179, 0.00174999]),
 'score_time': array([0.00242829, 0.00251007, 0.00212884, 0.00225592, 0.00324702,
        0.00429535, 0.00220299, 0.00215292, 0.00283003, 0.00229406]),
 'test_accuracy': array([0.69230769, 0.76923077, 0.53846154, 0.69230769, 0.38461538,
        0.69230769, 0.61538462, 0.75      , 0.66666667, 0.58333333]),
 'test_f1_macro': array([0.63888889, 0.70676692, 0.35      , 0.40909091, 0.27777778,
        0.56666667, 0.38095238, 0.62105263, 0.55555556, 0.36842105])}

In [46]:
scores_neigh = cross_validate(neigh, X_train, y_train, cv=10, scoring=['accuracy','f1_macro'])
scores_neigh

{'fit_time': array([0.00273514, 0.00221205, 0.00301003, 0.00186682, 0.00214624,
        0.00377989, 0.00446892, 0.00154996, 0.00234175, 0.00216413]),
 'score_time': array([0.00587392, 0.00515604, 0.0084188 , 0.00315094, 0.00717473,
        0.00474215, 0.0072    , 0.00336003, 0.00572014, 0.0041759 ]),
 'test_accuracy': array([0.61538462, 0.53846154, 0.53846154, 0.61538462, 0.53846154,
        0.61538462, 0.53846154, 0.58333333, 0.5       , 0.75      ]),
 'test_f1_macro': array([0.38095238, 0.35      , 0.35      , 0.38095238, 0.45833333,
        0.5112782 , 0.45833333, 0.36842105, 0.33333333, 0.42857143])}

In [47]:
scores_lr = cross_validate(lr, X_train, y_train, cv=10, scoring=['accuracy','f1_macro'])
scores_lr

{'fit_time': array([0.00603199, 0.00449896, 0.00404596, 0.00515485, 0.00381184,
        0.00431323, 0.00571489, 0.00399423, 0.00377083, 0.00399685]),
 'score_time': array([0.0022881 , 0.00229025, 0.00347018, 0.00218797, 0.00212407,
        0.00224185, 0.002321  , 0.00206804, 0.00203919, 0.002069  ]),
 'test_accuracy': array([0.69230769, 0.69230769, 0.61538462, 0.69230769, 0.69230769,
        0.69230769, 0.61538462, 0.75      , 0.75      , 0.75      ]),
 'test_f1_macro': array([0.40909091, 0.40909091, 0.38095238, 0.40909091, 0.40909091,
        0.40909091, 0.38095238, 0.42857143, 0.42857143, 0.42857143])}

## Επίδοση out-of-the-box

||Accuracy|F1-macro|
|---|--|----|
|**dummy**|0.7272|0.8421|
|**GNB**|0.5636|0.7|
|**KNN**|0.6545|0.7912|
|**LR**|0.7090|0.8397|